src/main/drivers/compass/compass_ist8310.c

```diff
@@ -0,0 +1,255 @@
+    // delay(5);
+    // write 0x01 to register 0x0A to set single measure mode
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL1, IST8310_ODR_SINGLE);
+    delay(5);
+
+    // Init setting : avg16 / pulse mode
+    ack = ack && busWriteRegister(dev, IST8310_REG_AVERAGE, IST8310_AVG_16);
+    delay(5);
+    ack = ack && busWriteRegister(dev, IST8310_REG_PDCNTL, IST8310_PULSE_DURATION_NORMAL);
+    delay(5);
+
+    // DRDY enable
+    ack = ack && busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp |= IST8310_CNTRL2_DRENA;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(5);
+
+    // DRDY polarity
+    ack = ack && busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp &= ~IST8310_CNTRL2_DRPOL;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(5);
+
+    return ack;
+}
+
+static bool ist8310Read(magDev_t * magDev, int16_t *magData)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    uint8_t buf[6];
+    // uint8_t LSB2FSV = 3; // 3mG - 14 bit
+    static uint8_t status;
+    static enum {
+        STATE_READ_INIT,
+        STATE_READ_STATUS,
+        STATE_WAIT_STATUS,
+        STATE_WAIT_READ,
+    } state = STATE_READ_INIT;
+
+    switch (state) {
+        default:
+        case STATE_READ_INIT:
+            magData[X] = 0;
+            magData[Y] = 0;
+            magData[Z] = 0;
+
+            state = STATE_READ_STATUS;
+            FALLTHROUGH;
+        case STATE_READ_STATUS:
+            busReadRegisterBufferStart(dev, IST8310_REG_STAT1, &status, sizeof(status));
+            state = STATE_WAIT_STATUS;
+            return false;
+        case STATE_WAIT_STATUS:
+            if ((status & IST8310_DRDY_MASK) == 0) {
+                state = STATE_READ_STATUS;
+                return false;
+            }
+
+            busReadRegisterBufferStart(dev, IST8310_REG_DATA, buf, sizeof(buf));
+            state = STATE_WAIT_READ;
+            return false;
+        case STATE_WAIT_READ:
+            // Looks like datasheet is incorrect and we need to invert Y axis to conform to right hand rule
+            // magData[X] =  (int16_t)(buf[1] << 8 | buf[0]) * LSB2FSV;
+            // magData[Y] = -(int16_t)(buf[3] << 8 | buf[2]) * LSB2FSV;
+            // magData[Z] =  (int16_t)(buf[5] << 8 | buf[4]) * LSB2FSV;
+
+            magData[X] = (int16_t)(buf[1] << 8 | buf[0]);
+            magData[Y] = (int16_t)(buf[3] << 8 | buf[2]);
+            magData[Z] = (int16_t)(buf[5] << 8 | buf[4]);
+
+            state = STATE_READ_STATUS;
+
+            return true;
+    }
+
+    // TODO: do cross axis compensation
+
+    return false;
+}
+
+static bool deviceDetect(magDev_t * magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    uint8_t sig = 0;
+    bool ack = busReadRegisterBuffer(dev, IST8310_REG_WHOAMI, &sig, 1);
+    ack = busReadRegisterBuffer(dev, IST8310_REG_WHOAMI, &sig, 1);
+    ack = busReadRegisterBuffer(dev, IST8310_REG_WHOAMI, &sig, 1);
```

Comment:
Why three calls exactly the same?

Reply:
See: https://github.com/liuheng135/drivers/blob/master/i2c/device/ist8310.c#L76-L95

Reply:
The detection seems to be working without this. I don't know why this code does it in this manner. I can do more tests tomorrow.

Reply:
Removed